### PR TITLE
llvm 11.0.0

### PR DESCRIPTION
**Pull request by kubkon**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by fxcoudert** [comment]
Can we add a short test that runs `llvm-config` and checks that it doesn't output things like `-llibxml2.tbd`? see https://github.com/Homebrew/homebrew-core/pull/62480

**Comment by fxcoudert** [comment]
`ccls` and `castxml` each need a revision bump (a separate commit for each, please)

**Comment by kubkon** [comment]
> `ccls` and `castxml` each need a revision bump (a separate commit for each, please)

@fxcoudert done in revised `133c386`. Lemme know if it makes sense though since I'm not very proficient in Ruby.

**Comment by jedisct1** [comment]
I still get the `-llibxml2.tbd` issue :(

**Comment by kubkon** [comment]
> I still get the `-llibxml2.tbd` issue :(

With LLVM 11? Huh, interesting cause I don't.

**Comment by jedisct1** [comment]
Yep, with LLVM 11, just compiled using your Formula. This is using Xcode 12.2 beta on Big Sur, though.

**Comment by kubkon** [comment]
> Yep, with LLVM 11, just compiled using your Formula. This is using Xcode 12.2 beta on Big Sur, though.

If the CI passes fine here, I reckon it'd be a good idea to inform upstream LLVM of the issue on Big Sur.

**Comment by fxcoudert** [comment]
> Regarding LLVM 11.0.0. They are aware about it AFAIK

If there is a patch, we can consider it. But we need to know that it works, that upstream is considering it.

**Comment by kubkon** [comment]
@jedisct1 @fxcoudert so it seems that my machine must be special, or perhaps I have something (mis)configured in a very special way; the `-llibxml2.tbd` regression is still there. What do you think the next steps should be? Should we merge this commenting out the test case and wait for the patch from upstream (after confirming with LLVM maintainers that it *will* happen first)?

**Comment by jedisct1** [comment]
The [patch](https://reviews.llvm.org/rGc3595d1069277b4ab0df49d7139b6f1bbc94f21c) is pretty short and simple.

If we can confirm that it works, maybe we should include it, no matter if this is the exact code that will eventually get merged upstream.

And we can always update the patch later if needed.

File: Formula/llvm.rb
Comment (kubkon) [inline Formula/llvm.rb:96]: @jedisct1 @fxcoudert a temp to see if the patch works. If it does, I'll go ahead and submit a PR to [Homebrew/formula-patches](https://github.com/Homebrew/formula-patches).

**Comment by SMillerDev** [comment]
We should really take the upstream patch unless it's impossible and/or upstream is unwilling to fix it.

File: Formula/llvm.rb
Comment (fxcoudert) [inline Formula/llvm.rb:314]: There should only be a single `test` block

File: Formula/llvm.rb
Comment (jedisct1) [reply Formula/llvm.rb]:
```text
==> Applying 11.0.0-llvm.diff
can't find file to patch at input line 5
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/llvm/lib/WindowsManifest/CMakeLists.txt b/llvm/lib/WindowsManifest/CMakeLists.txt
|index fe6ddcd414d5..8cfea1282ad6 100644
|--- a/llvm/lib/WindowsManifest/CMakeLists.txt
|+++ b/llvm/lib/WindowsManifest/CMakeLists.txt
--------------------------
No file to patch.  Skipping patch.
```

I'm not even sure that this is the right patch though. All these discussions and unrelated patches are quite confusing.

Comment (jedisct1) [reply Formula/llvm.rb]: The path should be `llvm-11.0.0.src/lib/WindowsManifest/CMakeLists.txt`.

The patch should also probably be reverted, not applied:

```diff
--- llvm-11.0.0.src/lib/WindowsManifest/CMakeLists.txt	2020-10-07 12:10:48.000000000 +0200
+++ llvm-11.0.0.src/lib/WindowsManifest/CMakeLists.txt	2020-10-13 21:16:16.000000000 +0200
@@ -8,16 +8,6 @@
 if(LIBXML2_LIBRARIES)
   target_link_libraries(LLVMWindowsManifest PUBLIC ${LIBXML2_LIBRARIES})
 
-  get_filename_component(xml2_library ${LIBXML2_LIBRARIES} NAME)
-  if (CMAKE_STATIC_LIBRARY_PREFIX AND
-      xml2_library MATCHES "^${CMAKE_STATIC_LIBRARY_PREFIX}.*${CMAKE_STATIC_LIBRARY_SUFFIX}$")
-    string(REGEX REPLACE "^${CMAKE_STATIC_LIBRARY_PREFIX}" "" xml2_library ${xml2_library})
-    string(REGEX REPLACE "${CMAKE_STATIC_LIBRARY_SUFFIX}$" "" xml2_library ${xml2_library})
-  elseif (CMAKE_SHARED_LIBRARY_PREFIX AND
-          xml2_library MATCHES "^${CMAKE_SHARED_LIBRARY_PREFIX}.*${CMAKE_SHARED_LIBRARY_SUFFIX}$")
-    string(REGEX REPLACE "^${CMAKE_SHARED_LIBRARY_PREFIX}" "" xml2_library ${xml2_library})
-    string(REGEX REPLACE "${CMAKE_SHARED_LIBRARY_SUFFIX}$" "" xml2_library ${xml2_library})
-  endif()
   set_property(TARGET LLVMWindowsManifest PROPERTY
-    LLVM_SYSTEM_LIBS ${xml2_library})
+    LLVM_SYSTEM_LIBS ${LIBXML2_LIBRARIES})
 endif()
```

I started a compile with it. Let's see how it goes.

Comment (jedisct1) [reply Formula/llvm.rb]: It works!

And fixes the issue we had with Zig.

Comment (kubkon) [reply Formula/llvm.rb]: Awesome! Thanks for taking the time to verify this and for all your help here @jedisct1!

Comment (jedisct1) [reply Formula/llvm.rb]: It actually picks up the same `tbd` file as before, but with the full path and without the extra `-l`.

Do we really want `-lxml2` instead? We can hardcode it, but this patch should restore the behavior from LLVM 10.0.0 that worked well.

Comment (kubkon) [reply Formula/llvm.rb]: No, I'd rather modify the added test case than force `-lxml2`. So with this patch, we are actually getting a fully qualified path to `libxml2.tbd`, did I get it right?

Comment (jedisct1) [reply Formula/llvm.rb]: Correct.

Comment (kubkon) [reply Formula/llvm.rb]: I've now tweaked the test to assert for this situation plus a situation where `llvm-config --system-libs` reports a fully qualified path to a `dylib` instead of `tbd` (as evident on my installation, for some reason).

**Comment by timmyjose** [comment]
I think special big thanks is due to the Homebrew team for working with the Zig folks so quickly and efficiently! It's quite heartening to see it in fact.

**Comment by fxcoudert** [comment]
```
==> Pouring openjdk-14.0.1.catalina.bottle.tar.gz
/usr/local/Cellar/openjdk/14.0.1/libexec/openjdk.jdk/Contents/MacOS/libjli.dylib: resource fork, Finder information, or similar detritus not allowed
Error: Failed changing dylib ID of /usr/local/Cellar/openjdk/14.0.1/libexec/openjdk.jdk/Contents/MacOS/libjli.dylib
  from @@HOMEBREW_PREFIX@@/opt/openjdk/libexec/openjdk.jdk/Contents/MacOS/libjli.dylib
    to /usr/local/opt/openjdk/libexec/openjdk.jdk/Contents/MacOS/libjli.dylib
Error: /usr/local/Cellar/openjdk/14.0.1/libexec/openjdk.jdk/Contents/MacOS/libjli.dylib: signing failed!
Warning: Bottle installation failed: building from source.
```

I've opened a brew issue: https://github.com/Homebrew/brew/issues/8922

**Comment by jedisct1** [comment]
Alternative one-liner patch that hardcodes `-lxml2`.

```diff
--- llvm-11.0.0.src/lib/WindowsManifest/CMakeLists.txt	2020-10-07 12:10:48.000000000 +0200
+++ llvm-11.0.0.src/lib/WindowsManifest/CMakeLists.txt	2020-10-14 18:31:13.000000000 +0200
@@ -19,5 +19,5 @@
     string(REGEX REPLACE "${CMAKE_SHARED_LIBRARY_SUFFIX}$" "" xml2_library ${xml2_library})
   endif()
   set_property(TARGET LLVMWindowsManifest PROPERTY
-    LLVM_SYSTEM_LIBS ${xml2_library})
+    LLVM_SYSTEM_LIBS "-lxml2")
 endif()
```

**Comment by kubkon** [comment]
> Alternative one-liner patch that hardcodes `-lxml2`.
> 
> ```diff
> --- llvm-11.0.0.src/lib/WindowsManifest/CMakeLists.txt	2020-10-07 12:10:48.000000000 +0200
> +++ llvm-11.0.0.src/lib/WindowsManifest/CMakeLists.txt	2020-10-14 18:31:13.000000000 +0200
> @@ -19,5 +19,5 @@
>      string(REGEX REPLACE "${CMAKE_SHARED_LIBRARY_SUFFIX}$" "" xml2_library ${xml2_library})
>    endif()
>    set_property(TARGET LLVMWindowsManifest PROPERTY
> -    LLVM_SYSTEM_LIBS ${xml2_library})
> +    LLVM_SYSTEM_LIBS "-lxml2")
>  endif()
> ```

As already consulted with @jedisct1 offline, the current patch submitted at Homebrew/formula-patches#309 doesn't really work 100% with the Zig project. Namely, building the stage2 compiler fails complaining that it has to link a fully qualified path to a dynamic library. With this in mind, I'd be keen to land @jedisct1's patch mentioned above as it is simple and easy to revert once this issue is fixed upstream. I'd be keen to learn everyone's opinion on what to do next.

cc @andrewrk this might be of interest to you plus I'd love to see what you reckon as well!

**Comment by fxcoudert** [comment]
Homebrew's position is: we typically only add patches that are already committed by upstream. So if you determine something is the "right fix", let the LLVM team (who knows their code best!) decide whether it's fine or not, and we'll ship it if they okay it.

**Comment by kubkon** [comment]
> Homebrew's position is: we typically only add patches that are already committed by upstream. So if you determine something is the "right fix", let the LLVM team (who knows their code best!) decide whether it's fine or not, and we'll ship it if they okay it.
> 
> We don't want to either risk shipping code that hasn't been validated by upstream, or risk becoming de facto maintainers :)

Roger that! @jedisct1 would you like to do the honours of submitting your fix to upstream?

**Comment by jedisct1** [comment]
This one-liner fixes the issue we have on macOS, but is probably not the correct fix to apply to other platforms :(

**Comment by bcardiff** [comment]
I noticed that llvm@10 will be dropped with this PR as is. The last content of Formula/llvm.rb should be copied to Formula/llvm@10.rb, right?

**Comment by fxcoudert** [comment]
> I noticed that llvm@10 will be dropped with this PR as is

Correct, unless we have other formulas that strictly depend on LLVM 10 (and won't build with 11), which does not look like it's the case. We do not intend to carry every version of the software, especially when it is not maintained anymore.

File: Formula/llvm.rb
Comment (bayandin) [inline Formula/llvm.rb:96]: If we have agreed to apply this patch, could you please create a PR to https://github.com/Homebrew/formula-patches (and change URL when it's merged).

File: Formula/llvm.rb
Comment (bayandin) [inline Formula/llvm.rb:328]:
```suggestion
    shell_output("#{bin}/llvm-config --system-libs").chomp.strip.split(" ").each do |lib|
      if lib.start_with?("-l")
        assert !lib.end_with?(".tbd"), "expected abs path when lib reported as .tbd"
        assert !lib.end_with?(".dylib"), "expected abs path when lib reported as .dylib"
      else
        p = Pathname.new(lib)
        if p.extname == ".tbd" || p.extname == ".dylib"
          assert p.absolute?, "expected abs path when lib reported as .tbd or .dylib"
        end
      end
    end
```

**Comment by kubkon** [comment]
Just an FYI that the formula now uses a patch which is officially applied upstream (massive thanks to @jedisct1 for coming up with it!). See Homebrew/formula-patches#310. To summarise, the patch included was applied to upstream in commit https://github.com/llvm/llvm-project/commit/c4d7536136b331bada079b2afbb2bd09ad8296bf, and resolves all our worries about improperly formatted output of `llvm-config --system-libs`, so after the tests pass I guess we are good to go!

**Comment by SMillerDev** [comment]
Please use the upstream commit as patch then rather than having homebrew host it.

**Comment by kubkon** [comment]
> Please use the upstream commit as patch then rather than having homebrew host it.

I tried doing that, however, the upstream patch is messy in the sense it affects files in other llvm subcomponents as well (namely, `lld` which we pull as a resource in the formula), plus I think some hunks/diffs in the upstream patch may rely on other patches that were applied to upstream which have not yet landed in `llvm-11.0.0`. @SMillerDev do you know if there's a way to select only a subset of a patch in homebrew with [`Formula.patch`](https://rubydoc.brew.sh/Formula#patch-class_method) class method? Essentially we'd like to extract only those bits listed in Homebrew/formula-patches#310.

**Comment by SMillerDev** [comment]
If you can find a way to do it using system commands I'd just download it as a resource.

**Comment by kubkon** [comment]
> If you can find a way to do it using system commands I'd just download it as a resource.

Cool, will give it a go, thanks!

**Comment by bcardiff** [comment]
Something that might be useful is to download the whole llvm-project instead of the individual components. That will allow using upstream commits and also simplify the patch creation. I think the formula download the subprojects individually as a consequence of evolution, I'm not 100% sure.

**Comment by tschoonj** [comment]
> Something that might be useful is to download the whole llvm-project instead of the individual components. That will allow using upstream commits and also simplify the patch creation. I think the formula download the subprojects individually as a consequence of evolution, I'm not 100% sure.

FWIW when this PR is merged in, I will be opening one that adds support for `flang`, the LLVM Fortran compile.. To do this I will need to switch to downloading the whole llvm-project, as one of the dependencies of `flang` is not available as a separate package, and is only available from within the llvm-project tarball.

**Comment by fxcoudert** [comment]
@tschoonj as I understand it flang is currently highly experimental, isn't it? I tried using it to assess it as a Fortran compiler for Apple Silicon but it seems pre-alpha at the moment.

**Comment by fxcoudert** [comment]
Right now it passes, it's an improvement over what we have, so I would suggest merging as-is and finesse it later :)
Formula-patches is made for situations like this, where the upstream patch does not apply cleanly.
@SMillerDev ?

**Comment by tschoonj** [comment]
> @tschoonj as I understand it flang is currently highly experimental, isn't it? I tried using it to assess it as a Fortran compiler for Apple Silicon but it seems pre-alpha at the moment.

It's officially part of the main llvm project as of 11.0, so my guess is that is considered stable now by upstream.

**Comment by kubkon** [comment]
> Right now it passes, it's an improvement over what we have, so I would suggest merging as-is and finesse it later :)
> Formula-patches is made for situations like this, where the upstream patch does not apply cleanly.
> @SMillerDev ?

I believe I got it working with only official gitrev's (still testing locally), so gimme a couple of minutes, perhaps we'll manage to do this without any patches stored at Homebrew at all. ;-)

**Comment by kubkon** [comment]
Oh and @tschoonj, already pulling `llvm-project` archive instead of multiple components independently ;-) Just need to wait for the build to finish for me locally, and the run the tests, and will push an update.

**Comment by fxcoudert** [comment]
I think the reason we were using different components is that it makes for a smaller download.

**Comment by kubkon** [comment]
> I think the reason we were using different components is that it makes for a smaller download.

Hmm, we should make a decision here then. If we go with one single download, I can make it work with official LLVM gitrevs. Otherwise, with multiple components, this becomes difficult due to the patches applied to multiple components at once. Thoughts?

**Comment by kubkon** [comment]
@fxcoudert Oh, and I should say I'm personally happy with either approach :-) If the latter works, I'll push it as a separate commit (avoiding rebases) so that if need be it's easy to take of.

**Comment by kubkon** [comment]
@fxcoudert And finally, the single download is manageable wrt to size. `llvm-project` archive is approx 80MB which should be manageable.

**Comment by kubkon** [comment]
@fxcoudert @SMillerDev @tschoonj fix pushed. Now we download only a single archive *and* we rely on unmodified upstream patches. Lemme know what you reckon guys!

**Comment by kubkon** [comment]
OK, can anyone offer some advice on triage failing with `full_index=1` param required please?

**Review by tschoonj — COMMENTED** [review]
Why was libcxxabi added to the runtimes?

**Comment by kubkon** [comment]
> Why was libcxxabi added to the runtimes?

What do you mean? The previous formula already did that however only for HEAD with a note to check when llvm-11 is released.

**Comment by tschoonj** [comment]
> > Why was libcxxabi added to the runtimes?
> 
> What do you mean? The previous formula already did that however only for HEAD with a note to check when llvm-11 is released.

Sorry, missed the note...

File: Formula/llvm.rb
Comment (jonchang) [inline Formula/llvm.rb:15]:
```suggestion
    patch do
      url "https://github.com/llvm/llvm-project/commit/c86f56e32e724c6018e579bb2bc11e667c96fc96.patch?full_index=1"
      sha256 "6e13e01b4f9037bb6f43f96cb752d23b367fe7db4b66d9bf2a4aeab9234b740a"
    end
```

* I think `:p1` is the default so we can omit that. (We should probably add an audit for that)
* `.patch` gives us a bit more info than the `.diff`, which helps us in case anyone forgets why the patch exists. (We should probably get https://github.com/Homebrew/brew/pull/8701 merged at some point)

File: Formula/llvm.rb
Comment (jonchang) [inline Formula/llvm.rb:114]: A companion pull request (after this is merged) to homebrew/formula-patches to delete this would be appreciated!

**Comment by kubkon** [comment]
Does anyone know if it is possible to apply a series of patches *after* all `resource` blocks are downloaded and extracted into the tree? This way I could leave the patches as they are while reverting back the formula to downloading all llvm components separately which should fix the CI.

**Comment by jonchang** [comment]
If you need to patch a resource I believe the patch block works inside the resource block too. Does that help?

**Comment by kubkon** [comment]
> If you need to patch a resource I believe the patch block works inside the resource block too. Does that help?

Unfortunately, it won't. The issue here is that we need to have all components in place before we apply the patches. In other words, first, we need to download `llvm-src`, `clang-src`, etc., extract them into one common dir, and *then* we can apply the patches since the patches are messy and touch upon multiple subprojects in one diff.

**Comment by kubkon** [comment]
OK, nvm, the hashes of diffs changed after I added `full_index=1`. Correcting now.

File: Formula/llvm.rb
Comment (kubkon) [reply Formula/llvm.rb]: Fixed!

**Comment by fxcoudert** [comment]
> downloading all llvm components separately which should fix the CI

What is the problem with CI and llvm-project?

**Comment by kubkon** [comment]
> > downloading all llvm components separately which should fix the CI
> 
> What is the problem with CI and llvm-project?

False alarm :-) I thought we hit the GitHub rate limit but it was my oversight with the hashes of the patches. It should be fixed now. Sorry for confusion!

File: Formula/llvm.rb
Comment (kubkon) [reply Formula/llvm.rb:114]: PR created: Homebrew/formula-patches#311

**Comment by kubkon** [comment]
Yay, everything passed! I reckon we're good to go if you agree! :-)

**Comment by jedisct1** [comment]
Woo-ooh!

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
Many thanks for the collective work on this one!

**Comment by lmurray** [comment]
> > I noticed that llvm@10 will be dropped with this PR as is
> 
> Correct, unless we have other formulas that strictly depend on LLVM 10 (and won't build with 11), which does not look like it's the case. We do not intend to carry every version of the software, especially when it is not maintained anymore.

I would like to chime in here to point out that the removal of llvm@10 has temporarily broken some users' development environments such as my own. Due to LLVM versions not being fully compatible with each other I pin my environments to specific versions of dependencies and then upgrade them when I have the available time to debug issues that occur due to the upgrade. Right now my environments are pinned by llvm@10 but it's no longer available, preventing me from installing a new environment on another system, or even rolling back if an upgrade to llvm@11 fails.

Currently llvm@9, llvm@8, llvm@7, and llvm@6 are already still available which means the removal of llvm@10 also goes against precedence.

I would like to propose that older versions of LLVM, and potentially other packages used by developers, are kept around at least for a few months so that the above issues are not encountered by users.

**Comment by lmurray** [comment]
Although I do admit that maybe I shouldn't be using Homebrew for this particular use case in the first place as it may be out of Homebrew's scope. Users that version pin could just use the official builds directly instead of going through Homebrew. This is what I do on other platforms anyway.

**Comment by fxcoudert** [comment]
Our criteria for versioned formulas are set out here: https://docs.brew.sh/Versions
If you believe LLVM 10 meets these criteria, feel free to open a pull request.

**Comment by lmurray** [comment]
Thank you for your response. The linked document clearly states that my use case of version pinning is not in Homebrew's scope which means there is no need to add LLVM 10 back.

**Comment by SMillerDev** [comment]
> Currently llvm@9, llvm@8, llvm@7, and llvm@6 are already still available which means the removal of llvm@10 also goes against precedence.

Rules change constantly based on new insights. Past results do not guarantee an outcome for the future in the case of Homebrew formulae.

**Pull request closed without merging**
